packages/vscode-extension/src/project/metro.ts

```diff
@@ -362,17 +388,21 @@ export class Metro implements Disposable {
       const description = newDebuggerPages[0].description;
       const isExpoGo = description === EXPO_GO_BUNDLE_ID || description === EXPO_GO_PACKAGE_NAME;
       if (isExpoGo) {
-        // Expo go apps using the new debugger would report at least two pages, first one being the Expo Go
-        // host runtime.
-        // If we detect Expo Go package (using description field), we want to pick the second page
-        // otherwise we pick the first one
-        // When only one page is listed, it means that the app runtime hasn't registered yet in which case
-        // we return undefined which triggers a retry mechanism.
-        if (newDebuggerPages.length === 1) {
-          return undefined;
-        } else {
-          return newDebuggerPages[1].webSocketDebuggerUrl;
+        // Expo go apps using the new debugger could report more then one page,
+        // if it exist the first one being the Expo Go host runtime.
+        // more over expo go on android has a bug causing newDebuggerPages
+        // from previously run applications to leek if the host application
+        // was not stopped.
+        // to solve both issues we check if the runtime is part of
+        // the host application process and select the last one that
+        // is not. To perform this check we use expo host functionality
+        // introduced in https://github.com/expo/expo/pull/32322/files
+        for (const newDebuggerPage of newDebuggerPages.reverse()) {
+          if (!(await this.isExpoHostRuntime(newDebuggerPage.webSocketDebuggerUrl))) {
```

Comment:
I don't think we only want to eliminate non-host runtimes here. Given we already connect to the websocket here, this method can actually verify if the runtime is active. This specifically makes sense given you write the old runtime can still be listed.

Since `iseExpoHostRuntime` returns `false` both in the case when we fail to connect with the runtime, and also in the case when we connected sucesfully to the runtime but it didn't have that flag set, I think we should actually check for the first case only. In the current form I'd assume that we'd end up selecting old runtime when the new one hasn't yet appeared on the list.